src/odoc.ml

```diff
@@ -452,28 +481,59 @@ module Gen (S : sig val sctx : SC.t end) = struct
     | Ok m -> m
     | Error (_, p1, p2) ->
       die "Package %s has two mld's with the same basename %s, %s"
-        (Package.Name.to_string pkg.Package.name)
+        (Package.Name.to_string pkg)
         (Path.to_string_maybe_quoted p1)
         (Path.to_string_maybe_quoted p2)
 
+  let setup_package_odoc_rules_def =
+    let module Input = struct
+      type t = Package.Name.t * Path.t list
+
+      let hash = Hashtbl.hash
+
+      let equal (x1, y1) (x2, y2) =
+        Package.Name.equal x1 x2 && List.equal Path.equal y1 y2
+
+      let to_sexp (name, paths) =
+        Dyn.Tuple
+          [ Package.Name.to_dyn name
+          ; Dyn.List (List.map ~f:Path.to_dyn paths)
+          ]
+        |> Dyn.to_sexp
+    end
+    in
+    Memo.create "setup-package-odoc-rules"
+      ~output:(Allow_cutoff (module Unit))
```

Comment:
Using `Unit` as the output type is not going to work in the long run (when we actually enable cross-run memoization), so we shouldn't get used to this idea.
That said, this case seems to be exactly what the "implicit output" for rules will cover, so it might be possible to keep this code basically unchanged.

Reply:
Should we leave a TODO or something for this? This would a problem in a few places.

Reply:
I guess an explicit TODO is not necessary because this just fits under a general "avoid side-effects inside memoized functions" umbrella, and currently annotating all such places with TODOs would be too much noise. Just something to keep in mind.